plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -819,6 +849,99 @@ private void finishOptimize(ConnectorSession session, IcebergTableExecuteHandle
         transaction = null;
     }
 
+    @Override
+    public void executeTableExecute(ConnectorSession session, ConnectorTableExecuteHandle tableExecuteHandle)
+    {
+        IcebergTableExecuteHandle executeHandle = (IcebergTableExecuteHandle) tableExecuteHandle;
+        switch (executeHandle.getProcedureId()) {
+            case VACUUM:
+                executeVacuum(session, executeHandle);
+                return;
+            default:
+                throw new IllegalArgumentException("Unknown procedure '" + executeHandle.getProcedureId() + "'");
+        }
+    }
+
+    private void executeVacuum(ConnectorSession session, IcebergTableExecuteHandle executeHandle)
+    {
+        IcebergVacuumHandle vacuumHandle = (IcebergVacuumHandle) executeHandle.getProcedureHandle();
+
+        Table table = catalog.loadTable(session, executeHandle.getSchemaTableName());
+        Map<String, String> properties = table.properties();
+        if (properties.containsKey(WRITE_LOCATION_PROVIDER_IMPL)) {
+            throw new TrinoException(NOT_SUPPORTED, "Table " + executeHandle.getSchemaTableName() + " specifies " + properties.get(WRITE_LOCATION_PROVIDER_IMPL) +
+                    " as a location provider. Writing to Iceberg tables with custom location provider is not supported.");
+        }
+        Duration retention = requireNonNull(vacuumHandle.getRetentionThreshold(), "retention is null");
+        Duration minRetention = requireNonNull(getVacuumMinRetention(session), "vacuumMinRetention is null");
+        checkArgument(retention.compareTo(
+                minRetention) >= 0,
+                "Retention specified (%s) is shorter than the minimum retention configured in the system (%s). " +
+                        "Minimum retention can be changed with %s configuration property or iceberg.%s session property",
+                retention,
+                minRetention,
+                IcebergConfig.VACUUM_MIN_RETENTION,
+                IcebergSessionProperties.VACUUM_MIN_RETENTION);
+
+        long expireTimestampMillis = System.currentTimeMillis() - retention.toMillis();
```

Comment:
nit: should we use `session.getStart()` rather than wall clock?

Reply:
It would be more future-proof in we turn out to run this on worker node some day - and there is time skew between workers and coordinator.

Reply:
I can change it